pkg/controller/serviceaccount/legacy_serviceaccount_token_cleaner.go

```diff
@@ -0,0 +1,232 @@
+	LegacySATokenCleanUpInterval time.Duration
+
+	// LegacySATokenCleanUpPeriod is the period of time since the last usage of an legacy token before it can be deleted.
+	LegacySATokenCleanUpPeriod time.Duration
+}
+
+// LegacySATokenCleaner is a controller that deletes legacy serviceaccount tokens that are not in use for a specified period of time.
+type LegacySATokenCleaner struct {
+	client clientset.Interface
+
+	saLister         listersv1.ServiceAccountLister
+	saInformerSynced cache.InformerSynced
+
+	secretLister         listersv1.SecretLister
+	secretInformerSynced cache.InformerSynced
+
+	podLister         listersv1.PodLister
+	podInformerSynced cache.InformerSynced
+
+	configMapLister         listersv1.ConfigMapLister
+	configMapInformerSynced cache.InformerSynced
+
+	legacySATokenCleanUpInterval time.Duration
+	legacySATokenCleanUpPeriod   time.Duration
+}
+
+// NewLegacySATokenCleaner returns a new *NewLegacySATokenCleaner.
+func NewLegacySATokenCleaner(saInformer coreinformers.ServiceAccountInformer, secretInformer coreinformers.SecretInformer, podInformer coreinformers.PodInformer, configMapInformer coreinformers.ConfigMapInformer, cl clientset.Interface, options LegacySATokenCleanerOptions) (*LegacySATokenCleaner, error) {
+	tc := &LegacySATokenCleaner{
+		client:                       cl,
+		saLister:                     saInformer.Lister(),
+		saInformerSynced:             saInformer.Informer().HasSynced,
+		secretLister:                 secretInformer.Lister(),
+		secretInformerSynced:         secretInformer.Informer().HasSynced,
+		configMapLister:              configMapInformer.Lister(),
+		configMapInformerSynced:      configMapInformer.Informer().HasSynced,
+		podLister:                    podInformer.Lister(),
+		podInformerSynced:            podInformer.Informer().HasSynced,
+		legacySATokenCleanUpPeriod:   options.LegacySATokenCleanUpPeriod,
+		legacySATokenCleanUpInterval: options.LegacySATokenCleanUpInterval,
+	}
+
+	return tc, nil
+}
+
+func (tc *LegacySATokenCleaner) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+
+	logger := klog.FromContext(ctx)
+	logger.Info("Starting legacy service account token cleaner controller")
+	defer logger.Info("Shutting down legacy service account token cleaner controller")
+
+	if !cache.WaitForNamedCacheSync("legacy-service-account-token-cleaner", ctx.Done(), tc.saInformerSynced, tc.secretInformerSynced, tc.configMapInformerSynced, tc.podInformerSynced) {
+		return
+	}
+
+	go wait.UntilWithContext(ctx, tc.evaluateSATokens, tc.legacySATokenCleanUpInterval)
+
+	<-ctx.Done()
+}
+
+func (tc *LegacySATokenCleaner) evaluateSATokens(ctx context.Context) {
+	logger := klog.FromContext(ctx)
+	secretList, err := tc.secretLister.Secrets(metav1.NamespaceAll).List(labels.Everything())
+	if err != nil {
+		logger.Error(err, "Getting secret list from secret cache")
+		return
+	}
+	for _, secret := range secretList {
```

Comment:
short-circuit on secrets of a type different from `kubernetes.io/service-account-token`

Reply:
Thank you, Stanislav. Fixed.